Add `dict` option to `nsolve`

Jayjayyy opened the pull request:
**Motivation:** If you're using `solve` with the `dict` option and want to call `nsolve` as a fallback, it would be useful, to get the same output return type and format. However, `nsolve` doesn't have a `dict` option, so far.

**Example:** Here's what my implementation looks like.
```python
>>> from sympy import *
>>> x, y = symbols("x y")
>>> equations = [Eq(2*x+y, 3), Eq(y-x, 1)]
>>> variables = [x, y]
>>> guesses = [1, 1]
>>> solve(equations, variables)
{x: 2/3, y: 5/3}
>>> solve(equations, variables, dict = True)
[{x: 2/3, y: 5/3}]
>>> nsolve(equations, variables, guesses)
Matrix([[0.666666666666667], [1.66666666666667]])
>>> nsolve(equations, variables, guesses, dict = True)
[{x: 0.666666666666667, y: 1.66666666666667}]
```

**Changes:** The `return` statement has been moved to the end of the function. It's now the same `return` for both one-dimensional and multi-dimensional calls. I didn't really change that much: those 20 deletions originate only from the necessary additional `else`. What was left to do was to distinguish between the different types, which the solution might have: either iterable (when solving multiple variables) or not (when solving only one variable). The most pythonic way to test if the result is iterable seems to be to assume the result to be iterable (line 2838) and to gracefully fail if not (line 2840).

Commit:
Update solvers.py

Add `dict` option to `nsolve`

asmeurer commented on sympy/solvers/solvers.py:
```diff
@@ -2804,26 +2804,42 @@ def nsolve(*args, **kwargs):
+    # check if user wants result to be returned as a dictionary
+    as_dict = kwargs.get('dict', False)
+    if as_dict:
+        # x won't be iterable (neither list, tuple nor Matrix) if there is only one variable in fargs, therefore exception TypeError will be raised
```
Can we check this condition directly instead of try/except?

asmeurer replied:
In fact, why not split this across the above if/else block (`if not isinstance(f, Matrix)`)?

Jayjayyy replied:
> Can we check this condition directly instead of try/except?

I got the impression, that try/except would be the best solution as it doesn't really matter if the return value is a Matrix or tuple or list or something else - it just depends on whether or not the result is iterable. This should be the only attribute which decides if the function needs to create the dictionary with `dict(zip(fargs, [sympify(xi) for xi in x]))` or `dict([(fargs, x)])`. Is that bad style?

> In fact, why not split this across the above if/else block

Sorry, I thought it would be cleaner (?) to complete the computation of the result first, and then to format the return value. I tried to make it similar to how the `solve` function handles it. Should I rewrite this with two `if as_dict:` statements at the end of each if/else block?

asmeurer replied:
We do have the `iterable` function (used above in this function), which handles the try/except logic. I don't like try/except when possible because it isn't as self-documenting, and it can mask unrelated errors if they happen to raise the same exception. 

But to me it seems simpler to just do the returns at the end of each block, where the results are calculated, because they already split the two possibilities (single equation or system). You can add `as_dict` to the top where the other options are gathered.

asmeurer replied:
(I don't feel strongly about doing it in the individual blocks or at the end. If you prefer the other way that's fine, but you should check `if isinstance(f, Matrix)` as above rather than use try/except).

asmeurer commented:
Add some tests.

Jayjayyy commented:
> Add some tests.

I'm sorry, how would I do this? I ran `sympy.doctest()` locally on my laptop but it took really long and quit with some errors, which had nothing to do with my changes or the solver functions at all, I believe.

asmeurer commented:
The doctests test the documentation examples (the actual tests are `sympy.test()`, which will take even longer to run on your laptop). Just add a function to sympy/solvers/tests/test_numeric.py. A simple test to test one single equation and one system, to make sure the dict option works should be fine. You can run just the the solver tests with `./bin/test sympy/solvers`, which is a lot faster than running the whole test suite.

asmeurer commented:
Also, if the doctests are failing for you please open an issue for that. The test suite should never fail on master.

Commit:
Update solvers.py

Commit:
Update test_numeric.py

Add tests for `dict` keyword argument in `nsolve`

Jayjayyy commented:
> But to me it seems simpler to just do the returns at the end of each block, where the results are calculated

Yes, I changed it.

> You can add `as_dict` to the top where the other options are gathered.

Done. I didn't see the consistency inside `nsolve` function at first, since `verbose` isn't assigned at the top, as well. But that should actually be at the top, too, right?

> Just add a function to sympy/solvers/tests/test_numeric.py. A simple test to test one single equation and one system, to make sure the dict option works should be fine.

I added something. Is that what you had in mind?

asmeurer commented:
The test looks good. One more thing that should be done is to document the option in the docstring. One thing that might be pointed out is that it uses a list for consistency with solve, even though nsolve (currently at least) only finds one solution at a time.

asmeurer commented on sympy/solvers/solvers.py:
```diff
@@ -2759,6 +2759,9 @@ def nsolve(*args, **kwargs):
         prec = None
 
+    # keyword argument to return result as a dictionary
+    as_dict = kwargs.get('dict', False)
```
Use `pop` instead of `get` so that it doesn't get passed to findroot.

Commit:
Update solvers.py

No need to pass `dict` option to `findroot`

Commit:
Update solvers.py

Add documentation for `dict` option to `nsolve` docstring

Jayjayyy commented:
> Use `pop`instead of `get`so that it doesn't get passed to findroot.

Of course, that makes sense. Thank you.

> One more thing that should be done is to document the option in the docstring

I added some lines. I'll happily take any suggestions for improvement. Should I add an example like `nsolve(x**2 - 2, 1, dict = True)`, too?

asmeurer reviewed (approved):
Looks good, thanks. Will merge once tests pass.

smichr commented:
I wonder if it might be nice to allow the symbol(s)/initial guess to be given as a dictionary, too. So instead of saying `.... [x, y], [1, 2]` you could say `..., {x:1 , y:2}`

Jayjayyy commented:
@asmeurer Thank you for your help. Sorry, if I'm slow at times - I'm not used to GitHub or large project such as SymPy, at all. However, I really enjoyed this. So thanks. :-)

@smichr Yes, I agree, that would be nice.

Also, I think there is a bug in `nsolve` which makes it reject overdetermined systems. `nsolve([x-1, 2*x-2], 0)` will raise `TypeError` even tough it would work fine with `findroot`. I noticed and fixed this while working on your proposed improvement to allow the symbol(s) and initial guess(es) given as a dictionary.

However, I might have some questions. What is the procedure in this case? Can I open a pull request even if my improvements for `nsolve` aren't complete, yet?

asmeurer commented:
I think it's fine that nsolve requires a list of symbols and initial guesses whenever passed a list of equations. It does work if you do that:

```
>>> nsolve([x-1, 2*x-2], [x], [0])
[1.0]
```

I don't think it should try to be too smart here.

Regarding PRs, yes feel free to open a pull request if you aren't finished yet. You can add `[WIP]` to the title to indicate that it is work in progress. That way people can start to review it early.
